non-fungible-token/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+		CannotDestroyClass,
+	}
+}
+
+pub type ClassInfoOf<T> =
+	ClassInfo<<T as Trait>::TokenId, <T as frame_system::Trait>::AccountId, <T as Trait>::ClassData>;
+pub type TokenInfoOf<T> = TokenInfo<<T as frame_system::Trait>::AccountId, <T as Trait>::TokenData>;
+
+decl_storage! {
+	trait Store for Module<T: Trait> as NonFungibleToken {
+		/// Next available class ID.
+		pub NextClassId get(fn next_class_id): T::ClassId;
+		/// Next available token ID.
+		pub NextTokenId get(fn next_token_id): T::TokenId;
+		/// Store class info.
+		///
+		/// Returns `None` if class info not set or removed.
+		pub Classes get(fn classes): map hasher(twox_64_concat) T::ClassId => Option<ClassInfoOf<T>>;
+		/// Store token info.
+		///
+		/// Returns `None` if token info not set or removed.
+		pub Tokens get(fn tokens): double_map hasher(twox_64_concat) T::ClassId, hasher(twox_64_concat) T::TokenId => Option<TokenInfoOf<T>>;
+		/// Token existence check by owner and class ID.
+		pub TokensByOwner get(fn tokens_by_owner): double_map hasher(twox_64_concat) T::AccountId, hasher(twox_64_concat) (T::ClassId, T::TokenId) => Option<()>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Create NFT(non fungible token) class
+	pub fn create_class(owner: &T::AccountId, metadata: CID, data: T::ClassData) -> Result<T::ClassId, DispatchError> {
+		let class_id = NextClassId::<T>::try_mutate(|id| -> Result<T::ClassId, DispatchError> {
+			let current_id = *id;
+			*id = id.checked_add(&One::one()).ok_or(Error::<T>::NoAvailableClassId)?;
+			Ok(current_id)
+		})?;
+
+		let info = ClassInfo {
+			metadata,
+			total_issuance: Default::default(),
+			owner: owner.clone(),
+			data,
+		};
+		Classes::<T>::insert(class_id, info);
+
+		Ok(class_id)
+	}
+
+	/// Transfer NFT(non fungible token) from `from` account to `to` account
+	pub fn transfer(from: &T::AccountId, to: &T::AccountId, token: (T::ClassId, T::TokenId)) -> DispatchResult {
+		if from == to {
+			return Ok(());
+		}
+
+		TokensByOwner::<T>::try_mutate_exists(from, token, |token_by_owner| -> DispatchResult {
+			ensure!(token_by_owner.take().is_some(), Error::<T>::NoPermission);
+			TokensByOwner::<T>::insert(to, token, ());
+			Ok(())
+		})?;
+
+		Tokens::<T>::try_mutate_exists(token.0, token.1, |token_info| -> DispatchResult {
+			let mut info = token_info.as_mut().ok_or(Error::<T>::TokenNotFound)?;
+			info.owner = to.clone();
+			Ok(())
+		})
+	}
+
+	/// Mint NFT(non fungible token) to `owner`
+	pub fn mint(
+		owner: &T::AccountId,
+		class_id: T::ClassId,
+		metadata: CID,
+		data: T::TokenData,
+	) -> Result<T::TokenId, DispatchError> {
+		let token_id = NextTokenId::<T>::try_mutate(|id| -> Result<T::TokenId, DispatchError> {
+			let current_id = *id;
+			*id = id.checked_add(&One::one()).ok_or(Error::<T>::NoAvailableTokenId)?;
+			Ok(current_id)
+		})?;
+
+		Classes::<T>::try_mutate(class_id, |class_info| -> DispatchResult {
```

Comment:
This could fail and NextTokenId still get updated.

Should put rest of the method into the `NextTokenId::try_mutate` closure